extensions/amp-story/1.0/amp-story.js

```diff
@@ -1258,17 +1319,43 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
     if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
+      pageIndex >= PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.PENDING
     ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
-        return Promise.resolve();
-      }
-      this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
+      return this.blockOnPendingSubscriptionsState_(targetPageId, direction);
+    }
 
-      // TODO(#37285): add SubscriptionService to actually trigger the subscription dialog.
+    // Navigation to the locked pages after the paywall page should be redirected to the paywall page.
+    if (
+      pageIndex > PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.BLOCKED
+    ) {
+      this.pageAfterGranted_ = targetPageId;
+      this.showSubscriptionsDialog_();
+      return this.switchTo_(
+        this.pages_[PAYWALL_PAGE_INDEX].element.id,
+        direction
+      );
+    }
+
+    if (
+      pageIndex === PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.BLOCKED
+    ) {
+      // Navigation to the paywall page should show dialog after delay.
+      this.paywallTimeout_ = setTimeout(() => {
+        this.pageAfterGranted_ = targetPageId;
+        this.showSubscriptionsDialog_();
+        this.paywallTimeout_ = null;
+      }, PAYWALL_DELAY_DURATION);
+    } else if (subscriptionsState !== SubscriptionsState.DISABLED) {
```

Comment:
Took me a few minutes to understand this condition and what it does, does it mean:

```suggestion
    if (pageIndex <= PAYWALL_PAGE_INDEX && StateProperty.SUBSCRIPTIONS_DIALOG_UI_STATE === true) {
```

?

If yes, that sounds like a much easier way to write it, + doesn't run unnecessary code